### DISCUSSION TITLE
Clarification around round-tripping KDL Numbers

### DISCUSSION THREAD

**Discussion by borland** [kdl-org/kdl#84] [Q&A] [answered]
KDL allows numbers to be specified either as integer, float/decimal, hex, octal, binary or scientific notation

The spec states

> There's 4 ways to represent numbers in KDL. KDL does not prescribe any representation for these numbers, and it's entirely up to individual implementations whether to represent all numbers with a single type, or to have different representations for different forms.

As an implementer, I take that to mean that internally I can represent all kinds of numbers using a 64 bit int, a byte array, or whatever internal format of my choosing.

However, when generating/outputting KDL, the spec doesn't say anything about number formats.

I'm not sure whether I need to preserve the number's input format, or whether I can "squash" it down to whatever format suits me. By example: When round-tripping this KDL document, do I need to output back the exact input of

```
ten 10 0xA 1e1 1E+1 0b1010 10.0
```

Or can I simply output this?

```
ten 10 10 10 10 10 10
```

If the KDL language definition is fine with the latter, could you please add a note to the spec about round-tripping numbers and that this is not expected?

Thanks

**Comment by borland** [comment]
Aside: my personal opinion is that while it’s probably ok to squash things like 0xA and 1e1 down to “10”, a well behaved implementation shouldn’t squash a floating point/decimal number down into an integer, as that represents a different data type and might be significant. 

IE: 10.0 should not squash down to 10, but other squashing seems fair game

**Comment by zkat** [comment, marked as answer]
You can output numbers any way you want. Implementations for people who actually care about formatting would need to be document-oriented and have extra data structures or options that tell the KDL handler how to actually write out the data. Does that make sense? That means outputting `ten 10 10 10 10 10 10` is totally fine. It's the same *value*. The various input syntaxes are only different in syntax, not semantics.

**Reply by alilleybrinker** [reply]
Relatedly, I would take this to mean that, when reading `ten 10 10 10 10 10 10` as input, you're entirely fine to decide "I want this to be floaty" and treat is as a float and do float things to it, because `10` here represents a particular _value_ but not a particular _type_. Do I have that right?

**Reply by zkat** [reply]
correct. Numbers in KDL have no specified implementation, and have multiple types of input formats on top of that. It is entirely up to individual implementations how to represent them, regardless of whether they have decimal components or not. The idea is that, if you want to constrain this, you use a KDL Schema (https://github.com/kdl-org/kdl/blob/main/SCHEMA-SPEC.md)

**Reply by alilleybrinker** [reply]
Excellent! Thanks for the clarification! :smile:
